crates/accelerate/src/sparse_observable/mod.rs

```diff
@@ -1026,6 +1233,181 @@ impl ::std::ops::Neg for SparseObservable {
     }
 }
 
+/// Worker objects for the [SparseObservable::compose]-alike functions.
+mod compose {
+    use super::*;
+
+    /// An non-scalar entry in the multi-Cartesian product iteration.
+    ///
+    /// Each [MultipleItem] corresponds to a bit index that becomes a summation as part of the
+    /// composition, so the multi-Cartesian product has to keep iterating through it.
+    struct MultipleItem {
+        /// The index into the full `bit_terms` [Vec] that this item refers to.
+        loc: usize,
+        /// The next item in the slice that should be written out.  If this equal to the length
+        /// of the slice, the implication is that it needs to be reset to 0 by a lower-index
+        /// item getting incremented and flowing forwards.
+        cur: usize,
+        /// The underlying slice of the multiple iteration, from the compose lookup table.
+        slice: &'static [(Complex64, BitTerm)],
+    }
+
+    /// An implementation of the multiple Cartesian-product iterator that produces the sum terms
+    /// that make up the composition of two individual [SparseObservable] terms.
+    ///
+    /// This mutates itself in-place to keep track of the state, which allows us to re-use
+    /// partial results and to avoid allocations per item (since we have to copy out of the
+    /// buffers to the output [SparseObservable] each time anyway).
+    pub struct Iter {
+        num_qubits: u32,
+        /// Tracking data for the places in the output that have multiple branches to take.
+        multiples: Vec<MultipleItem>,
+        /// Stack of the coefficients to this point.  We could recalculate by a full
+        /// multiplication on each go, but most steps will be in the low indices, where we can
+        /// re-use all the multiplications that came before.  This is one longer than the length
+        /// of `multliples`, because it starts off populated with the product of the
+        /// non-multiple coefficients (or 1).
+        coeffs: Vec<Complex64>,
+        /// The full set of indices (including ones that don't correspond to multiples).  Within
+        /// a given iteration, this never changes; it's just stored to make it easier to memcpy
+        /// out of.
+        indices: Vec<u32>,
+        /// The full set of [BitTerm]s (including ones that don't correspond to multiples).  The
+        /// multiple ones get updated inplace during iteration.
+        bit_terms: Vec<BitTerm>,
+        /// Whether iteration has started.
+        started: bool,
+        /// We always have at least one item to put out, but we're exhausted once we've iterated
+        /// through all our product terms.
```

Comment:
Might be nice to explicitly something like this, I was a little confused when reading the case where the `matmul` is None and `exhausted` was directly set 🙂 
```suggestion
        /// We always have at least one item to put out. We're exhausted once we've iterated
        /// through all our product terms, or the product is zero.
```

Reply:
Oh, the comment is actually just wrong - we don't always have at least one term to put out.  Changed in 5dd0666c99.